tests/functional/base.py

```diff
@@ -0,0 +1,662 @@
+        for _id, container in containers.items():
+            if container['name'] == name:
+                raise AssertionError('Container: %s is running' % name)
+        return True
+
+    def get_initial_state(self):
+        self._containers = self._get_containers(all=True)
+
+    def _get_containers(self, all=False):
+        cmd = ['docker', 'ps']
+        if all:
+            cmd.append('-a')
+        output = subprocess.check_output(cmd)
+        _containers = OrderedDict()
+
+        for line in output.splitlines()[1:]:
+            container = self._get_container(line)
+            _containers[container['id']] = container
+        return _containers
+
+    def _get_container(self, line):
+        words = re.split(' {2,}', line)
+        if len(words) == 6:
+            words = words[:-1] + [''] + words[-1:]
+        container_id, image, command, created, status, ports, names = words
+        return {
+            'id': container_id,
+            'image': image,
+            'command': command,
+            'created': created,
+            'status': status,
+            'ports': ports,
+            'names': names
+        }
+
+
+class KubernetesProviderTestSuite(BaseProviderTestSuite):
+    """
+    Base test suite for Kubernetes.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        cls.disable_selinux()
+        logger.debug('setUpClass...')
+        logger.debug('Stopping existing kubernetes instance, if any...')
+        kubernetes.stop()
+        logger.debug('Starting kubernetes instance...')
+        kubernetes.start()
+        time.sleep(10)
+        cls.answers = anymarkup.parse(kubernetes.answers(), 'ini')
+
+    @classmethod
+    def tearDownClass(cls):
+        kubernetes.stop()
+        cls.enable_selinux()
+
+    def tearDown(self):
+        logger.debug('Teardown ...')
+        pods = self._get_pods()
+        services = self._get_services()
+        rcs = self._get_rcs()
+
+        # clean up newly created pods
+        logger.debug('clean up pods')
+        for pod in pods:
+            if pod not in self._pods:
+                subprocess.check_output('kubectl delete pod ' + pod, shell=True)
+
+        # clean up newly created services
+        for service in services:
+            if service not in self._services:
+                subprocess.check_output('kubectl delete service ' + service, shell=True)
+
+        # clean up newly created rcs
+        for rc in rcs:
+            if rc not in self._rcs:
+                subprocess.check_output('kubectl delete rc ' + rc, shell=True)
+
+        for pod in pods:
+            if pod not in self._pods:
+                self.assertPod(pod, exists=False, timeout=360)
+
+        for service in services:
+            if service not in self._services:
+                self.assertService(service, exists=False, timeout=360)
+
+        for rc in rcs:
+            if rc not in self._rcs:
+                self.assertRc(rc, exists=False, timeout=360)
+
```

Comment:
No need for this many for statements, have a look at the stop command under providers/kubernetes.sh script for a way to stop without listing each service, pod, etc.